obp/ope/estimators_multi.py

```diff
@@ -0,0 +1,1806 @@
+
+    estimator_name: str, default='multi_weighted_ipw'.
+        Name of the estimator.
+
+    References
+    ------------
+    Aman Agarwal, Soumya Basu, Tobias Schnabel, and Thorsten Joachims.
+    "Effective Evaluation using Logged Bandit Feedback from Multiple Loggers.", 2018.
+
+    Nathan Kallus, Yuta Saito, and Masatoshi Uehara.
+    "Optimal Off-Policy Evaluation from Multiple Logging Policies.", 2021.
+
+    """
+
+    estimator_name: str = "multi_weighted_ipw"
+
+    def _estimate_round_rewards(
+        self,
+        reward: np.ndarray,
+        action: np.ndarray,
+        pscore: np.ndarray,
+        stratum_idx: np.ndarray,
+        action_dist: np.ndarray,
+        position: Optional[np.ndarray] = None,
+        **kwargs,
+    ) -> np.ndarray:
+        """Estimate round-wise (or sample-wise) rewards.
+
+        Parameters
+        ----------
+        reward: array-like, shape (n_rounds,)
+            Rewards observed for each data in logged bandit data, i.e., :math:`r_i`.
+
+        action: array-like, shape (n_rounds,)
+            Actions sampled by the logging/behavior policy for each data in logged bandit data, i.e., :math:`a_i`.
+
+        pscore: array-like, shape (n_rounds,)
+            Action choice probabilities of the logging/behavior policy (propensity scores), i.e., :math:`\\pi_k(a_i|x_i)`.
+            If `use_estimated_pscore` is False, `pscore` must be given.
+
+        stratum_idx: array-like, shape (n_rounds,)
+            Indices to differentiate the logging/behavior policy that generate each data, i.e., :math:`k`.
+
+        action_dist: array-like, shape (n_rounds, n_actions, len_list)
+            Action choice probabilities of the evaluation policy (can be deterministic), i.e., :math:`\\pi_e(a_i|x_i)`.
+
+        position: array-like, shape (n_rounds,), default=None
+            Indices to differentiate positions in a recommendation interface where the actions are presented.
+            If None, the effect of position on the reward will be ignored.
+            (If only a single action is chosen for each data, you can just ignore this argument.)
+
+        Returns
+        ----------
+        estimated_rewards: array-like, shape (n_rounds,)
+            Estimated rewards for each observation.
+
+        """
+        if position is None:
+            position = np.zeros(action_dist.shape[0], dtype=int)
+
+        n = action.shape[0]
+        iw = action_dist[np.arange(n), action, position] / pscore
+        # weight clipping
+        if isinstance(iw, np.ndarray):
+            iw = np.minimum(iw, self.lambda_)
+
+        unique_stratum_idx, n_data_strata = np.unique(stratum_idx, return_counts=True)
+        var_k = np.zeros(unique_stratum_idx.shape[0])
+        for k in unique_stratum_idx:
+            idx_ = stratum_idx == k
+            var_k[k] = np.var(reward[idx_] * iw[idx_])
+        weight_k = n / (var_k * np.sum(n_data_strata / var_k))
```

Comment:
[ask]
Is this a typo?

```suggestion
        weight_k = n_data_strata / (var_k * np.sum(n_data_strata / var_k))
```

Reply:
@fullflu Thanks! Actually, I think my implementation is correct. In the following definition, there is no (1/n), but in our implementation, we always take the average of the estimated rewards in `estimate_policy_value`. So, we have to multiply the estimated_rewards by `n` in advance for this estimator.

<img width="1022" alt="スクリーンショット 2022-02-09 2 29 51" src="https://user-images.githubusercontent.com/32621707/153142632-03e17f70-e857-4d74-8c85-878d2fe75d76.png">

Reply:
OK, I agree with you!!